### PR TITLE
[SPARK-42360][SQL] Transform LeftOuter join with IsNull filter on right side to Anti join

### PR DESCRIPTION
**Pull request by wankunde**
### What changes were proposed in this pull request?

Optimize left outer join to anti join if exists IsNull filter on the right side.
For example, query
```sql
SELECT t1.*
FROM t1 LEFT JOIN t2
ON t1.id = t2.id
WHERE t2.id is null and t1.id = 1 and Rand() < 0.5
```
can be optimized to
```sql
SELECT t1.*
FROM t1 LEFT ANTI JOIN t2
ON t1.id = t2.id
WHERE t1.id = 1 and Rand() < 0.5
```

PostgreSQL also has this optimization
image

### Why are the changes needed?

Optimize left outer join.

Local benchmark code:
```
object LeftOuterToAntiJoinBenchmark extends SqlBasedBenchmark {

  override def runBenchmarkSuite(mainArgs: Array[String]): Unit = {
    runBenchmark("Compare left outer join to anti join") {
      spark.range(1, 100000, 1, 5)
        .selectExpr("id%5 as id")
        .createOrReplaceTempView("tab")
      val testQuery =
        s"""SELECT t1.*
           |FROM tab t1
           |LEFT OUTER JOIN tab t2
           |ON t1.id = t2.id
           |WHERE t2.id is null
           |""".stripMargin
      val benchmark =
        new Benchmark("Compare left outer join to anti join", 100000, output = output)
      benchmark.addCase(s"Left outer join result") { _ =>
        withSQLConf(SQLConf.OPTIMIZER_EXCLUDED_RULES.key ->
          "org.apache.spark.sql.catalyst.optimizer.EliminateOuterJoin") {
          val res = spark.sql(testQuery)
          res.noop()
        }
      }
      benchmark.addCase(s"Anti join result") { _ =>
        val res = spark.sql(testQuery)
        res.noop()
      }
      benchmark.run()
    }
  }
}
```
benchmark result
```
Intel(R) Core(TM) i9-9980HK CPU @ 2.40GHz
Compare left outer join to anti join:     Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------------------------------
Left outer join result                             9500           9884         544          0.0       94997.6       1.0X
Anti join result                                    287            356          80          0.3        2869.9      33.1X
```

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

Added UT

### CODE REVIEW

File: sql/core/src/test/resources/tpcds-plan-stability/approved-plans-v2_7/q78.sf100/simplified.txt [hunk 1]
```diff
@@ -17,36 +17,35 @@ TakeOrderedAndProject [ss_sold_year,ss_item_sk,ss_customer_sk,ss_qty,ss_wc,ss_sp
...
-                                      Filter [sr_ticket_number]
-                                        SortMergeJoin [ss_ticket_number,ss_item_sk,sr_ticket_number,sr_item_sk]
-                                          InputAdapter
-                                            WholeStageCodegen (2)
-                                              Sort [ss_ticket_number,ss_item_sk]
-                                                InputAdapter
-                                                  Exchange [ss_ticket_number,ss_item_sk] #2
-                                                    WholeStageCodegen (1)
-                                                      Filter [ss_item_sk,ss_customer_sk]
+                                      SortMergeJoin [ss_ticket_number,ss_item_sk,sr_ticket_number,sr_item_sk]
```
Comment (wangyum) [inline sql/core/src/test/resources/tpcds-plan-stability/approved-plans-v2_7/q78.sf100/simplified.txt:20]: Do you have the benchmark number if active this change?

Comment (wankunde) [reply sql/core/src/test/resources/tpcds-plan-stability/approved-plans-v2_7/q78.sf100/simplified.txt:20]: Add a local benchmark code and result in PR description.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/joins.scala [hunk 2]
```diff
@@ -202,8 +202,42 @@ object EliminateOuterJoin extends Rule[LogicalPlan] with PredicateHelper {
...
+        }
+
+      case _ => None
+    }
+  }
+
   def apply(plan: LogicalPlan): LogicalPlan = plan.transformWithPruning(
     _.containsPattern(OUTER_JOIN), ruleId) {
+    case LeftOuterJoinsWithNullFilter(newProject) =>
+      newProject
```
Comment (wangyum) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/joins.scala:240]: Makes it more clear?
```scala
    case p @ Project(projectList, Filter(cond, j @ Join(left, right, LeftOuter, _, _)))
        if projectList.forall(canEvaluate(_, left)) =>
      val notNullAttrs =
        AttributeSet(right.constraints.collect { case IsNotNull(a: Attribute) => a })
      if (notNullAttrs.nonEmpty) {
        val (isNullConditions, rest) = splitConjunctivePredicates(cond).partition {
          case IsNull(a: Attribute) => notNullAttrs.contains(a)
          case _ => false
        }
        if (isNullConditions.nonEmpty) {
          val newJoin = j.copy(joinType = LeftAnti)
          p.copy(child = rest.reduceLeftOption(And).map(Filter(_, newJoin)).getOrElse(newJoin))
        } else {
          p
        }
      } else {
        p
      }
```

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/OuterJoinEliminationSuite.scala [hunk 3]
```diff
@@ -315,4 +315,28 @@ class OuterJoinEliminationSuite extends PlanTest {
...
+    val x = testRelation.subquery("x")
+    val y = testRelation1.subquery("y")
+    comparePlans(InferFilterOptimize.execute(
+      x.join(y, LeftOuter, Some($"a" === $"d"))
+        .where($"d".isNull && rand(1) < 0.5)
+        .select($"a", $"b", $"c").analyze),
+      x.join(y.where($"d".isNotNull), LeftAnti, Some($"a" === $"d"))
+        .where(rand(1) < 0.5)
+        .select($"a", $"b", $"c").analyze
+    )
```
Comment (wangyum) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/OuterJoinEliminationSuite.scala:340]: We can remove `InferFiltersFromConstraints`:
```scala
    val x = testRelation.subquery("x")
    val y = testRelation1.subquery("y").where($"d".isNotNull)

    val originQuery = x.join(y, LeftOuter, Some($"a" === $"d"))
      .where($"d".isNull && rand(1) < 0.5)
      .select($"a", $"b", $"c")

    val correctAnswer = x.join(y, LeftAnti, Some($"a" === $"d"))
      .where(rand(1) < 0.5)
      .select($"a", $"b", $"c")

    comparePlans(Optimize.execute(originQuery.analyze), correctAnswer.analyze)
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/joins.scala [hunk 4]
```diff
@@ -204,6 +204,26 @@ object EliminateOuterJoin extends Rule[LogicalPlan] with PredicateHelper {
...
+        if projectList.forall(canEvaluate(_, left)) =>
+      val notNullAttrs =
+        AttributeSet(right.constraints.collect { case IsNotNull(a: Attribute) => a })
+      if (notNullAttrs.nonEmpty) {
+        val (isNullConditions, rest) = splitConjunctivePredicates(cond).partition {
+          case IsNull(a: Attribute) => notNullAttrs.contains(a)
+          case _ => false
+        }
+        if (isNullConditions.nonEmpty &&
+          rest.forall(_.references.intersect(right.outputSet).isEmpty)) {
```
Comment (wangyum) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/joins.scala:217]: Please add more test cases to cover the different code paths.

Comment (wankunde) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/joins.scala:217]: Thanks @wangyum , I have updated the UT.

Another question, for this UT,  could we optimize `testRelation1.subquery("y").where($"d".isNotNull).where($"d".isNull && $"f" > 10)` to `EmptyRelation` to skip data scan?

**Comment by wankunde** [comment]
Retest this please

**Comment by wankunde** [comment]
cc @cloud-fan

**Pull request closed without merging**